state/watcher.go

```diff
@@ -496,6 +496,33 @@ func watchApplicationRelations(backend modelBackend, applicationName string) Str
 	return newRelationLifeSuspendedWatcher(backend, members, filter, nil)
 }
 
+// WatchModelMachinesLifeAndAgentStartTimes watches the machines in the model
+// for changes to the Life or AgentStartTime fields.
+func (st *State) WatchModelMachinesLifeAndAgentStartTimes() StringsWatcher {
+	docStates := make(map[string]string)
+	return newCollectionWatcher(st, colWCfg{
+		col: machinesC,
+		filter: func(id interface{}) bool {
+			m, err := st.Machine(id.(string))
+			if err != nil {
+				return false
+			}
+
+			// Ignore containers
+			if m.doc.ContainerType != "" {
+				return false
+			}
+
+			mState := fmt.Sprintf("%s-%d-%d", m.doc.DocID, m.doc.Life, m.doc.AgentStartedAt.Unix())
```

Comment:
This would be better as a struct - I don't think you need docID in there though?

Reply:
You do because we are watching the collection so we need to distinguish between individual records. We could alternatively use ID (the machine ID) which is shorter. However, given that I will change this to use batch mode it is probably easier to use the docID (which we already know) plus the output of the DB query (Life, AgentStartedAt) to reduce the number of fields we pull out of the DB.